src/Lorentz/Contracts/RegistryDAO/Types.hs

```diff
@@ -0,0 +1,105 @@
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+-- | RegistryDAO Types
+module Lorentz.Contracts.RegistryDAO.Types
+  ( RegistryDaoContractExtra (..)
+  , RegistryEntry (..)
+  , RegistryUpdate (..)
+  , RegistryDaoProposalMetadata (..)
+  , IsoRegistryDaoProposalMetadata
+  ) where
+
+import Lorentz
+
+import qualified Lorentz.Contracts.BaseDAO.Types as DAO
+
+
+data RegistryDaoContractExtra k v = RegistryDaoContractExtra
+  { ceRegistry :: BigMap k (RegistryEntry v)
+  }
+  deriving stock (Generic)
+
+instance (NiceComparable k, Ord k, IsoValue k, IsoValue v)
+  => (IsoValue (RegistryDaoContractExtra k v))
+
+instance (HasAnnotation k, HasAnnotation v)
+  => HasAnnotation (RegistryDaoContractExtra k v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (Ord k, TypeHasDoc k, TypeHasDoc v)
+  => TypeHasDoc (RegistryDaoContractExtra k v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly2TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryDaoContractExtra ByteString ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryDaoContractExtra ByteString ByteString)
+
+instance (Ord k) => Default (RegistryDaoContractExtra k v) where
+  def = RegistryDaoContractExtra mempty
+
+
+data RegistryEntry v = RegistryEntry
+  { reValue :: Maybe v
+   -- optional statistics on given proposal
+  , reLastAgoraPostId :: Natural
+  , reLastUpdated :: Timestamp
+  } deriving stock (Generic)
+  deriving anyclass (IsoValue)
+
+instance (HasAnnotation v)
+  => HasAnnotation (RegistryEntry v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (TypeHasDoc v)
+  => TypeHasDoc (RegistryEntry v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly1TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryEntry ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryEntry ByteString)
+
+
+data RegistryDaoProposalMetadata k v = RegistryDaoProposalMetadata
+  { pmAgoraPostId :: Natural
+  , pmDiff :: [RegistryUpdate k v]
+  }
+  deriving stock (Generic)
+
+type IsoRegistryDaoProposalMetadata k v = (NiceComparable k, Ord k, IsoValue k, IsoValue v)
+
+instance IsoRegistryDaoProposalMetadata k v
+  => (IsoValue (RegistryDaoProposalMetadata k v))
+
+instance (HasAnnotation k, HasAnnotation v)
+  => HasAnnotation (RegistryDaoProposalMetadata k v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (TypeHasDoc k, TypeHasDoc v)
+  => TypeHasDoc (RegistryDaoProposalMetadata k v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly2TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryDaoProposalMetadata ByteString ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryDaoProposalMetadata ByteString ByteString)
+
+data RegistryUpdate k v = RegistryUpdateItem
+  { ruKey :: k
+  , ruNewValue :: Maybe v
+  , ruOldValue :: Maybe v -- why we need oldValue
```

Comment:
Okay, I think in some cases it can really matter indeed. I don't like `Maybe v` because it forces you to specify the old value (I assume `Nothing` here means "value is absent" rather than "I don't care"), but sometimes you just want to forcefully update a value (maybe delete it). So I think it should be `Maybe (Maybe v)`, but it looked ugly, so I decided not to include it :yao:

Maybe let's not include for now, but suggest it as additional feature to be added in the next version?

Reply:
For me, I'm good with not including this for now.